File: zenodo_code_package/src/python/par2_cli.py
#!/usr/bin/env python3
"""
PAR(2) Discovery Engine - Command Line Interface

Usage: python par2_cli.py --input data.csv --period 24
"""

import argparse
import json
import sys
from pathlib import Path
from datetime import datetime

import numpy as np
import pandas as pd

from par2_core import run_par2_analysis, run_batch_analysis


DEFAULT_CLOCK_GENES = ['Per2', 'Arntl', 'Clock', 'Per1', 'Cry1', 'Cry2', 'Nr1d1', 'Nr1d2']
DEFAULT_TARGET_GENES = ['Wee1', 'Ccnd1', 'Myc', 'Cdk1', 'Tp53']


def parse_args():
    parser = argparse.ArgumentParser(
        description='PAR(2) Discovery Engine - Phase-Amplitude-Relationship Analysis',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Input Format:
  CSV with first column = gene names, subsequent columns = timepoints
  Header row should contain timepoint labels (e.g., CT0, CT4, CT8...)

Example:
  python par2_cli.py -i expression.csv -p 24 -c Per2,Arntl -t Wee1,Ccnd1
        """
    )
    
    parser.add_argument('-i', '--input', required=True, help='Input CSV file with gene expression data')
    parser.add_argument('-p', '--period', type=float, default=24.0, help='Circadian period in hours (default: 24)')
    parser.add_argument('-c', '--clock', default=','.join(DEFAULT_CLOCK_GENES), help='Comma-separated list of clock genes')
    parser.add_argument('-t', '--target', default=','.join(DEFAULT_TARGET_GENES), help='Comma-separated list of target genes')
    parser.add_argument('-o', '--output', help='Output JSON file (default: stdout)')
    parser.add_argument('--threshold', type=float, default=0.05, help='Significance threshold (default: 0.05)')
    
    return parser.parse_args()


def parse_csv(filepath: str) -> tuple:
    """Parse CSV file with gene expression data"""
    df = pd.read_csv(filepath, index_col=0)
    
    # Parse timepoints from column headers
    timepoints = []
    for col in df.columns:
        import re
        match = re.search(r'(\d+)', str(col))
        if match:
            timepoints.append(float(match.group(1)))
        else:
            timepoints.append(len(timepoints) * 4)  # Default 4-hour intervals
    
    genes = {}
    for gene_name in df.index:
        expression = df.loc[gene_name].values.astype(float)
        if not np.any(np.isnan(expression)):
            genes[str(gene_name).strip('"')] = expression.tolist()
    
    return genes, timepoints


def main():
    args = parse_args()
    
    if not Path(args.input).exists():
        print(f"Error: Input file not found: {args.input}", file=sys.stderr)
        sys.exit(1)
    
    print(f"PAR(2) Discovery Engine", file=sys.stderr)
    print(f"Input: {args.input}", file=sys.stderr)
    print(f"Period: {args.period}h", file=sys.stderr)
    print("", file=sys.stderr)
    
    genes, timepoints = parse_csv(args.input)
    print(f"Loaded {len(genes)} genes, {len(timepoints)} timepoints", file=sys.stderr)
    
    clock_genes = [g.strip() for g in args.clock.split(',')]
    target_genes = [g.strip() for g in args.target.split(',')]
    
    # Find available genes
    available_clocks = [g for g in clock_genes if g in genes]
    available_targets = [g for g in target_genes if g in genes]
    
    if not available_clocks:
        print("Error: No clock genes found in data", file=sys.stderr)
        print(f"Looking for: {', '.join(clock_genes)}", file=sys.stderr)
        print(f"Available: {', '.join(list(genes.keys())[:20])}...", file=sys.stderr)
        sys.exit(1)
    
    if not available_targets:
        print("Error: No target genes found in data", file=sys.stderr)
        sys.exit(1)
    
    print(f"Clock genes found: {', '.join(available_clocks)}", file=sys.stderr)
    print(f"Target genes found: {', '.join(available_targets)}", file=sys.stderr)
    print("", file=sys.stderr)
    
    # Build analysis list
    analyses = []
    for clock in available_clocks:
        for target in available_targets:
            analyses.append({
                'id': f'{clock}->{target}',
                'clockData': {'time': timepoints, 'expression': genes[clock]},
                'targetData': {'time': timepoints, 'expression': genes[target]}
            })
    
    print(f"Running {len(analyses)} pair analyses...", file=sys.stderr)
    
    batch_result = run_batch_analysis(analyses, args.period, args.threshold)
    
    # Format output
    output = {
        'timestamp': datetime.now().isoformat(),
        'config': {
            'period': args.period,
            'threshold': args.threshold,
            'inputFile': Path(args.input).name
        },
        'summary': {
            'totalPairs': len(analyses),
            'significantBeforeFDR': batch_result['fdrCorrection']['significantCount'],
            'significantAfterFDR': batch_result['fdrCorrection']['significantCountAfterFDR']
        },
        'results': [
            {'pair': analyses[i]['id'], **batch_result['results'][i]}
            for i in range(len(analyses))
            if batch_result['results'][i]['significant']
        ]
    }
    
    json_output = json.dumps(output, indent=2)
    
    if args.output:
        with open(args.output, 'w') as f:
            f.write(json_output)
        print(f"Results written to {args.output}", file=sys.stderr)
    else:
        print(json_output)
    
    print("", file=sys.stderr)
    print(f"Summary: {output['summary']['significantAfterFDR']}/{output['summary']['totalPairs']} pairs significant (FDR < {args.threshold})", file=sys.stderr)


if __name__ == '__main__':
    main()

File: zenodo_code_package/src/typescript/par2-cli.ts
#!/usr/bin/env npx tsx
/**
 * PAR(2) Discovery Engine - Command Line Interface
 * 
 * Usage: npx tsx par2-cli.ts --input data.csv --period 24
 */

import * as fs from 'fs';
import * as path from 'path';
import { runPAR2Analysis, runBatchAnalysis, GeneData } from './par2-core';

interface CLIOptions {
  input: string;
  period: number;
  clockGenes: string[];
  targetGenes: string[];
  output?: string;
  threshold: number;
}

function parseArgs(): CLIOptions {
  const args = process.argv.slice(2);
  const options: CLIOptions = {
    input: '',
    period: 24,
    clockGenes: ['Per2', 'Arntl', 'Clock', 'Per1', 'Cry1', 'Cry2', 'Nr1d1', 'Nr1d2'],
    targetGenes: ['Wee1', 'Ccnd1', 'Myc', 'Cdk1', 'Tp53'],
    threshold: 0.05
  };

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '--input':
      case '-i':
        options.input = args[++i];
        break;
      case '--period':
      case '-p':
        options.period = parseFloat(args[++i]);
        break;
      case '--clock':
      case '-c':
        options.clockGenes = args[++i].split(',');
        break;
      case '--target':
      case '-t':
        options.targetGenes = args[++i].split(',');
        break;
      case '--output':
      case '-o':
        options.output = args[++i];
        break;
      case '--threshold':
        options.threshold = parseFloat(args[++i]);
        break;
      case '--help':
      case '-h':
        printHelp();
        process.exit(0);
    }
  }

  return options;
}

function printHelp(): void {
  console.log(`
PAR(2) Discovery Engine - Command Line Interface

Usage: npx tsx par2-cli.ts [options]

Options:
  -i, --input <file>      Input CSV file with gene expression data (required)
  -p, --period <hours>    Circadian period in hours (default: 24)
  -c, --clock <genes>     Comma-separated list of clock genes
  -t, --target <genes>    Comma-separated list of target genes
  -o, --output <file>     Output JSON file (default: stdout)
  --threshold <value>     Significance threshold (default: 0.05)
  -h, --help              Show this help message

Input Format:
  CSV with first column = gene names, subsequent columns = timepoints
  Header row should contain timepoint labels (e.g., CT0, CT4, CT8...)

Example:
  npx tsx par2-cli.ts -i expression.csv -p 24 -c Per2,Arntl -t Wee1,Ccnd1
`);
}

function parseCSV(filepath: string): { genes: Record<string, number[]>; timepoints: number[] } {
  const content = fs.readFileSync(filepath, 'utf-8');
  const lines = content.trim().split('\n');
  
  if (lines.length < 2) {
    throw new Error('CSV must have at least a header row and one data row');
  }

  const header = lines[0].split(',').map(h => h.trim());
  
  // Parse timepoints from header (skip first column which is gene name)
  const timepoints: number[] = [];
  for (let i = 1; i < header.length; i++) {
    const match = header[i].match(/(\d+)/);
    if (match) {
      timepoints.push(parseFloat(match[1]));
    } else {
      timepoints.push((i - 1) * 4); // Default 4-hour intervals
    }
  }

  const genes: Record<string, number[]> = {};
  
  for (let i = 1; i < lines.length; i++) {
    const cols = lines[i].split(',').map(c => c.trim());
    if (cols.length < 2) continue;
    
    const geneName = cols[0].replace(/"/g, '');
    const expression = cols.slice(1).map(v => parseFloat(v)).filter(v => !isNaN(v));
    
    if (expression.length > 0) {
      genes[geneName] = expression;
    }
  }

  return { genes, timepoints };
}

function main(): void {
  const options = parseArgs();

  if (!options.input) {
    console.error('Error: Input file is required. Use --help for usage.');
    process.exit(1);
  }

  if (!fs.existsSync(options.input)) {
    console.error(`Error: Input file not found: ${options.input}`);
    process.exit(1);
  }

  console.error(`PAR(2) Discovery Engine`);
  console.error(`Input: ${options.input}`);
  console.error(`Period: ${options.period}h`);
  console.error('');

  const { genes, timepoints } = parseCSV(options.input);
  console.error(`Loaded ${Object.keys(genes).length} genes, ${timepoints.length} timepoints`);

  // Find available clock and target genes
  const availableClocks = options.clockGenes.filter(g => genes[g]);
  const availableTargets = options.targetGenes.filter(g => genes[g]);

  if (availableClocks.length === 0) {
    console.error('Error: No clock genes found in data');
    console.error(`Looking for: ${options.clockGenes.join(', ')}`);
    console.error(`Available: ${Object.keys(genes).slice(0, 20).join(', ')}...`);
    process.exit(1);
  }

  if (availableTargets.length === 0) {
    console.error('Error: No target genes found in data');
    process.exit(1);
  }

  console.error(`Clock genes found: ${availableClocks.join(', ')}`);
  console.error(`Target genes found: ${availableTargets.join(', ')}`);
  console.error('');

  // Run batch analysis
  const analyses: { targetData: GeneData; clockData: GeneData; id: string }[] = [];

  for (const clock of availableClocks) {
    for (const target of availableTargets) {
      analyses.push({
        id: `${clock}->${target}`,
        clockData: { time: timepoints, expression: genes[clock] },
        targetData: { time: timepoints, expression: genes[target] }
      });
    }
  }

  console.error(`Running ${analyses.length} pair analyses...`);

  const batchResult = runBatchAnalysis(analyses, {
    period: options.period,
    significanceThreshold: options.threshold
  });

  // Format output
  const output = {
    timestamp: new Date().toISOString(),
    config: {
      period: options.period,
      threshold: options.threshold,
      inputFile: path.basename(options.input)
    },
    summary: {
      totalPairs: analyses.length,
      significantBeforeFDR: batchResult.fdrCorrection.significantCount,
      significantAfterFDR: batchResult.fdrCorrection.significantCountAfterFDR
    },
    results: analyses.map((a, i) => ({
      pair: a.id,
      ...batchResult.results[i]
    })).filter(r => r.significant)
  };

  const jsonOutput = JSON.stringify(output, null, 2);

  if (options.output) {
    fs.writeFileSync(options.output, jsonOutput);
    console.error(`Results written to ${options.output}`);
  } else {
    console.log(jsonOutput);
  }

  console.error('');
  console.error(`Summary: ${output.summary.significantAfterFDR}/${output.summary.totalPairs} pairs significant (FDR < ${options.threshold})`);
}

main();
